modules/core/src/main/java/org/apache/ignite/internal/util/GridHandleTable.java

```diff
@@ -107,6 +121,14 @@ public int lookup(Object obj) {
      * Resets table to its initial (empty) state.
      */
     public void clear() {
+        if (size < objs.length) {
+            if (shrink()) {
+                size = 0;
+
+                return;
+            }
+        }
+
         System.arraycopy(spineEmpty, 0, spine, 0, spineEmpty.length);
         System.arraycopy(nextEmpty, 0, next, 0, nextEmpty.length);
```

Comment:
How about clearing not all nextEmpty and objs array, but only first size elements?

Reply:
Sounds legit

Reply:
On the other hand, objs should be cleared entirely to avoid referencing objects that are no longer needed

Reply:
'Objs' array is filling from [0 to size), all the rest elements are always null.

Reply:
Yes, you're right, I had a little mixup)